File: solutions/2302-count-subarrays-with-score-less-than-k.py
# 2302. Count Subarrays With Score Less Than K
# Topics: Array, Binary Search, Sliding Window, Prefix Sum
# https://leetcode.com/problems/count-subarrays-with-score-less-than-k/


from typing import List

class Solution:
    def countSubarrays(self, nums: List[int], k: int) -> int:
        # Time: O(n)
        # Space: O(1)
        num_subarrays = 0
        window_sum = 0
        left_index = 0

        # Variable sliding window
        for right_index in range(len(nums)):
            window_sum += nums[right_index]
            while left_index <= right_index and window_sum * (right_index - left_index + 1) >= k:
                window_sum -= nums[left_index]
                left_index += 1

            num_subarrays += right_index - left_index + 1

        return num_subarrays

File: solutions/2302-count-subarrays-with-score-less-than-k.cpp
// 2302. Count Subarrays With Score Less Than K
// Topics: Array, Binary Search, Sliding Window, Prefix Sum
// https://leetcode.com/problems/count-subarrays-with-score-less-than-k/


class Solution {
    public:
        long long countSubarrays(vector<int>& nums, long long k) {
            // Time: O(n)
            // Space: O(1)
            long long numSubarrays = 0;
            long long windowSum = 0;
            int leftIndex = 0;
    
            // Variable Sliding Window
            for (int rightIndex = 0; rightIndex < nums.size(); ++rightIndex) {
                windowSum += nums[rightIndex];
    
                while (leftIndex <= rightIndex && windowSum * (rightIndex - leftIndex + 1) >= k)
                    windowSum -= nums[leftIndex++];
    
                numSubarrays += rightIndex - leftIndex + 1;
            }
    
            return numSubarrays;
        }
    };